### ISSUE TITLE
Make m2_name and M2 metadata attributes?

### ISSUE THREAD

**Issue by dkahle** [musicman3320/m2r#115]
The report has pointed out to me something that feels a little strange: that M2 objects that have analogous data structures in R don't get parsed into those data structures. Is that a good idea? Example (updated for clarity):
```r
# make a matrix in M2 from R
(mat. <- m2_matrix.(matrix(c(1,2,3,4,5,6), nrow = 3, ncol = 2)))
# Starting M2
# M2 Pointer Object
#   ExternalString : map((ZZ)^3,(ZZ)^2,{{1, 4}, {2, 5}, {3, 6}})
#          M2 Name : m2rintmatrix00000001
#         M2 Class : Matrix (Type)


# look at the structure
str(mat.)
# List of 4
#  $ ext_str       : chr "map((ZZ)^3,(ZZ)^2,{{1, 4}, {2, 5}, {3, 6}})"
#  $ m2_name       : chr "m2rintmatrix00000001"
#  $ m2_class      : chr "Matrix"
#  $ m2_class_class: chr "Type"
#  - attr(*, "class")= chr [1:2] "m2_pointer" "m2"


# pull that matrix back into R
m2_parse(mat.)
# $m2_name
# [1] "m2rintmatrix00000001"
# 
# $rmatrix
#      [,1] [,2]
# [1,]    1    4
# [2,]    2    5
# [3,]    3    6
# 
# $ring
# M2 Ring: ZZ[], grevlex order
# attr(,"class")
# [1] "m2_matrix" "m2"  
```

As far as `m2_parse()` is concerned, it seems obvious to me that the resulting data structure should know that the object is an M2 object, that the object has a binding in M2, and what that binding is.  It should also know other relevant info (e.g. the referent ring) related to the primary object (e.g. the matrix). Obviously, `m2_parse()` currently does all this, but the behavior can be somewhat unexpected: when I `m2_parse()` a matrix, I expect to get a matrix that I can play with, but I don't: I get a list that contains that matrix.

I think it may make more sense to hold the `m2_name` and the other info (like `ring`) as attributes, likely as a list.  For example, something like 
```r
m2_parse(mat.)
#      [,1] [,2]
# [1,]    1    4
# [2,]    2    5
# [3,]    3    6
# attr(,"class")
# [1] "m2_matrix" "m2"  
# attr(,"m2_name")
# [1] "m2rintmatrix00000001"
# attr(,"m2_meta")
# attr(,"m2_meta")$ring
# M2 Ring: ZZ[], grevlex order
```
(Here `m2_meta` is [an attribute (metadata)](http://adv-r.had.co.nz/Data-structures.html#attributes) that contains a named list.) 

This isn't always going to work.  For example, the above example seems to make sense when the corresponding ring is the integers, because R can represent arrays of integers, but it won't work for `mpoly`s, because while R has a data structure for arrays, it doesn't have one for arrays of `mpoly`s (barring our hack allowing it to do so).  On the other hand, arrays of `mpoly`s can be implemented, and otherwise R has a many of the things we'd want: ints, complexes, etc.

Another thought is this: `m2_name` is all over the place in __m2r__ as an element of a list, not an attribute.  So, that'd be some work to change, but I'd imagine it's not that bad, and we can easily create an accessor that allows us to grab it, for example:
```r
get_m2_name <- function (x) {
  if ( is.m2(x) ) {
    attr(x, "m2_name")
  } else {
    character(0)
  }
}
```

Thoughts?

**Comment by musicman3320** [comment]
To me, this makes sense for data that is common among ALL `m2` objects, such as `m2_name`.  However, anything else would be hard to distinguish in a predictable manner.  Even `ring` attribute is not universal: some types do not have that attribute, and others (in the future) may have 2 or more.  More generally, deciding what properties are "attributes" and which are "members" seems counter-intuitive.  For instance, which properties of `m2_ring` should be attributes?  It is unclear to me how to do this in a consistently across all objects.  

Note: philosophically, this is sort of the point of calling the non-dotted version of a function.  Internally, a non-dotted function should call the dotted version, pass the result to `m2_parse()` then extract what it needs and return.  Perhaps `m2_matrix()` should do this, and we should abandon the `m2_matrix` type?  Is this currently an issue with any types other than `m2_matrix`?

**Comment by musicman3320** [comment]
Another potential issue: right now a matrix only has `ring` and `m2_name` in addition to the actual matrix, but what if we want to add other properties later, things that wouldn't fit as attributes? That would require changing `m2_matrix` BACK to a list, which could break code.  As it is currently written, all we would have to do is add another named entry, which does not break existing code.  

Perhaps I am just not familiar with how attributes work/are treated in R.

**Comment by dkahle** [comment]
Attributes can be anything in the R language, so I think we'd be fine there.

For the first point, I think `m2_name` will be the only thing common to all `m2` objects, and that's why I include it as a "first level" attribute (that's not a technical term, btw).  So I would see all `m2` objects having three attributes: their class, which they already have, `m2_name`, and then `m2_metadata`, which would be any metadata associated with the object as a named list (possibly empty `list()`), and whose contents would change according to the type of `m2` object.

Rings seem to present basic examples of "hard" objects: objects with no analogues in R.  Maybe they would just be a `NA` (missing value of type `logical(1)`; you can't assign attributes to `NULL`, which might seem preferable).  For instance, here is currently what we get with `ring()`:
```r
(my_ring <- ring(c("x","y"), coefring = "ZZ"))
# M2 Ring: ZZ[x,y], grevlex order

str(my_ring)
# $ coefring: chr "ZZ"
# $ vars    : chr [1:2] "x" "y"
# $ order   : chr "grevlex"
# - attr(*, "class")= chr [1:2] "m2_polynomialring" "m2"
```
But maybe the structure should be
```r
my_ring <- NA

# set all attributes at once
attributes(my_ring) <- list(
  class = c("m2_ring", "m2"),
  m2_name = "m2rintring00000001",
  m2_metadata = list(
    coefring = "ZZ",
    vars = c("x","y"),
    order = "grevlex"
  )
)

str(my_ring)
# Classes 'm2_ring', 'm2'  atomic [1:1] NA
#   ..- attr(*, "m2_name")= chr "m2rintring00000001"
#   ..- attr(*, "m2_metadata")=List of 3
#   .. ..$ coefring: chr "ZZ"
#   .. ..$ vars    : chr [1:2] "x" "y"
#   .. ..$ order   : chr "grevlex"
```
In other words, a `m2_ring` object (or objects that don't have analogues in R), don't need to have a primary object associated with them, because they aren't used in R for anything.  Does that make sense? Ones that do assume that data type (e.g. integer matrices), but still contain metadata.

**Comment by musicman3320** [comment]
Actually, this is a rather slick way to do things then.  Essentially, EVERYTHING is an attribute EXCEPT the fundamental data, if it has a natural R counterpart.  I really like this idea.  

Is it too late to make this change before the CRAN submission?  This will require a LOT of code modifications, but better to do it before the first submission so as not to break anyone's code.

**Comment by musicman3320** [comment]
It does seem a bit more of a pain to access an attribute (`attr(x, "m2_metadata")$ring` instead of `x$ring`).  Is there a shorthand way to do it?  Or perhaps we can override one for `m2_metadata` specifically?

**Comment by dkahle** [comment]
It's not too late, no. I think you're right about the number of modifications, but I would guess they're easy to find/make.  Obviously, we can expect some bumps along the way.  I agree with the cumbersome call – `x$ring` is way easier than `attr(x, "m2_metadata")$ring`, hm. 

We could so something like `m2_attr(x, "ring")`:
```r
m2_attr <- function (x) {
  if(!is.m2(x)) return(NULL)
  attr(x, "m2_metadata")[[x, exact = TRUE]]
}
```
Thoughts? Is there a better name for the function?, or an easier way that you can think of?

**Comment by dkahle** [comment]
(In the above, maybe we'd change the `m2_metadata` attribute to `m2_attr`.)

**Comment by musicman3320** [comment]
This is much shorter.  I like it.  How do we want to split the work for this?  I likely can't touch the code until tomorrow evening.

**Comment by dkahle** [comment]
Once you do the parser stuff, I think I can probably clean the rest.

**Comment by musicman3320** [comment]
Just added `m2_name()` and `m2_meta()` to m2_utility.R.  Could you add/modify what is needed to be able to do 
```
m2_name(x) <- "m2rring00000001"
m2_attr(x, "ring") <- myring
```
just as `class()` and `attr()` currently allow?

**Comment by dkahle** [comment]
You should be good to go now.

**Comment by musicman3320** [comment]
Just committed big changes.  All properties are now stored as attributes.  All unit tests are passing.  

In addition to the functions discussed above, the following utility function makes it easier to create m2 objects. 
```
m2_structure(
  m2_name = "ZZ",
  m2_class = "m2_polynomialring",
  m2_meta = list(
    coefring = "ZZ",
    vars = NULL,
    order = "grevlex"
  )
)
```
